cogs/lastactive.py

```diff
@@ -0,0 +1,126 @@
+"""
+lastactive (Ren v2) - Tracks when a user was last active.
+
+How to use:
+    bot.last_active[server_id][channel_id][author_id] -> Retrieves datetime (UTC) of last message
+    sent by that author in that channel in that server.
+
+Last updated by jangarong on August 22nd, 2019.
+"""
+import os
+import asyncio
+import json
+from datetime import datetime, timedelta
+
+
+class LastActive:
+
+    def __init__(self, bot, fromJson=False, toJson=True, limit=500):
+        """
+        fromJson : (Boolean) If true, it will retrieve last active data from json based on path. If
+            false, it will retrieve last active data via logs from channels.
+        toJson : (Boolean) If true, it will save the json file in the desired path periodically.
+        limit : (Integer) Number of messages read in the chat history of each channel. Keep in mind
+            that the larger the number, the longer it will take for the bot to set up.
+
+        To change these values, see the setup function down below.
+        """
+        self.bot = bot
+        self.bot.lastActive = {}
+        self.fromJson = fromJson
+        self.toJson = toJson
+        self.limit = limit
+
+        # working directory = cogs
+        self.jsonPath = os.path.abspath(os.path.dirname(__file__))[:-len('/cogs')] + \
+                        '/data/lastactive/last_active.json'
+
+        # create loop that goes every 15 minutes
+        self.lastChecked = datetime.now() - timedelta(minutes=15)
+        self.bgTask = self.bot.loop.create_task(self.jsonLoop())
+
+    def createFolder(self):
+        """Creates a folder in case if one did not exist already."""
+        folder_name = self.jsonPath[:-len('last_active.json')]
+        if not os.path.exists(folder_name):
+            os.makedirs(folder_name)
+
+    def dumpJson(self):
+        """Saves dictionary into json."""
+        self.createFolder()
+        with open(self.jsonPath, 'w') as f:
+            json.dump(self.bot.lastActive, f)
+
+    def loadJson(self):
+        """Loads dictionary into json."""
+        with open(self.jsonPath, 'r') as f:
+            self.bot.lastActive = json.load(f)
+
+    # iterate per day to save last active data
+    async def jsonLoop(self):
+        """Saves dictionary into json file every 15 minutes."""
+        while self == self.bot.get_cog("LastActive"):
+            if (self.lastChecked + timedelta(minutes=15)).minute == datetime.now().minute:
```

Comment:
That should definitely be a `<=` instead of `==` for the time comparison. If something happens and this code runs 16 minutes later it will never save

Reply:
That is true. However I changed the loop so that it no longer uses an if statement (since it will be dumping to jsondt every 60 seconds), and just uses the delay.